File: apps/api/app/services/exams.py
"""Examens certifiants (spec §5.5, contrat Phase 2 §2).

- Fichiers `CONTENT_DIR/<pack>/exams/*.json` (schéma `content/schema/exam.schema.json`), lus à la demande ;
  un dossier absent ou un fichier invalide n'empêche pas les autres examens d'être servis.
- Ouvert quand le test d'unité (`kind: unit_test`) de chaque unité requise est **réussi** (meilleur score ≥ 0,7)
  ou que l'unité a été sautée grâce au placement (contrat parcours §2).
- Médias (contrat parcours §1) : un item oral sans référence F0 et un item d'écoute tonal sans audio natif ne sont
  pas notés (retirés du dénominateur) ; une compétence sans item noté a un score null et est exclue de la règle
  « chaque compétence ≥ 0,5 ». Moins de 15 items notés : examen indisponible (`media_missing`).
- Une tentative soumise verrouille l'examen `retryAfterHours` ; une tentative expire après
  `durationMinutes` + 2 min de grâce.
- Le serveur note : il reconstruit chaque exercice avec la graine de la tentative (identique au client)
  et rejoue `evaluate` — portage de `buildExam`/`gradeExam` (packages/core/src/exams.ts) : leçon synthétique
  d'id `examId` dont l'étape k est le k-ième item tous items confondus (ordre des sections), donc graine
  d'un item `${attemptId}:${examId}:${flatIndex}`.
"""

import json
import logging
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import ExamAttempt
from app.services import progression
from app.services.content import Pack, concept_documents
from app.services.engine import (
    SPEAK_PASS_SCORE,
    ContentError,
    Exercise,
    build_exercise,
    content_index,
    evaluate,
    lesson_concepts_for_units,
)
from app.services.media import exam_item_graded

logger = logging.getLogger(__name__)

SKILLS = ("listening", "reading", "vocabulary", "speaking")
SPEECH_TYPES = frozenset({"speak_repeat", "tone_produce"})
GRACE = timedelta(minutes=2)
MIN_SKILL_SCORE = 0.5
# En dessous, l'examen est « indisponible pour le moment » (médias manquants).
MIN_GRADED_ITEMS = 15
MEDIA_MISSING = "media_missing"
# Tolérance de comparaison flottante (EPSILON de exams.ts).
EPSILON = 1e-9


@dataclass(frozen=True)
class ExamSpec:
    id: str
    level: str
    certificate: dict[str, str]
    requires_units: list[str]
    duration_minutes: int
    pass_threshold: float
    retry_after_hours: int
    sections: list[dict[str, Any]]
    raw: dict[str, Any]

    def items(self) -> list[tuple[str, int]]:
        return [(s["skill"], i) for s in self.sections for i in range(len(s["items"]))]

    def steps(self, skill: str) -> list[dict[str, Any]]:
        section = next((s for s in self.sections if s["skill"] == skill), None)
        return [item["step"] for item in section["items"]] if section else []


def parse_exam(raw: dict[str, Any]) -> ExamSpec:
    """Lecture tolérante mais stricte sur ce que la notation utilise ; ValueError si inexploitable."""
    try:
        sections = raw["sections"]
        for section in sections:
            if section["skill"] not in SKILLS:
                raise ValueError(f"compétence inconnue : {section['skill']}")
            for item in section["items"]:
                if not isinstance(item.get("step"), dict) or "type" not in item["step"]:
                    raise ValueError("item sans step")
        return ExamSpec(
            id=str(raw["id"]),
            level=str(raw["level"]),
            certificate=dict(raw["certificate"]),
            requires_units=list(raw["requiresUnits"]),
            duration_minutes=int(raw["durationMinutes"]),
            pass_threshold=float(raw["passThreshold"]),
            retry_after_hours=int(raw.get("retryAfterHours", 48)),
            sections=list(sections),
            raw=raw,
        )
    except (KeyError, TypeError) as exc:
        raise ValueError(f"examen invalide : {exc!r}") from exc


def load_exams(pack: Pack) -> dict[str, ExamSpec]:
    """Examens du pack, triés par niveau. Fichiers invalides ignorés (journalisés)."""
    directory = pack.directory / "exams"
    exams: dict[str, ExamSpec] = {}
    if not directory.is_dir():
        return exams
    for path in sorted(directory.glob("*.json")):
        try:
            exam = parse_exam(json.loads(path.read_text(encoding="utf-8")))
        except (OSError, ValueError) as exc:
            logger.warning("Examen ignoré %s : %s", path, exc)
            continue
        exams[exam.id] = exam
    return dict(sorted(exams.items(), key=lambda kv: kv[1].level))


# --- Ouverture et verrou --------------------------------------------------------------------


def unit_test_lessons(pack: Pack, unit_id: str) -> list[str]:
    """Tests d'unité (`kind: unit_test`) ; à défaut toutes les leçons de l'unité ; unité vide → liste vide."""
    unit = next((u for u in pack.units if u.id == unit_id), None)
    if unit is None:
        return []
    tests = progression.unit_test_lessons(pack, unit)
    return tests or [lid for lid in unit.lessons if lid in pack.lessons]


def is_unlocked(db: Session, user_id: str, pack: Pack, exam: ExamSpec) -> bool:
    """Tests d'unité réussis (≥ 0,7) ou unités sautées grâce au placement, pour chaque unité requise."""
    from app.services.learner import progress_state  # import local : learner dépend du planificateur

    if not exam.requires_units:
        return False
    return progression.done_units(pack, progress_state(db, user_id, pack), exam.requires_units)


def last_submitted(db: Session, user_id: str, exam_id: str) -> ExamAttempt | None:
    return db.scalar(
        select(ExamAttempt)
        .where(ExamAttempt.user_id == user_id, ExamAttempt.exam_id == exam_id, ExamAttempt.submitted_at.is_not(None))
        .order_by(ExamAttempt.submitted_at.desc())
        .limit(1)
    )


def next_attempt_at(exam: ExamSpec, last: ExamAttempt | None, now: datetime) -> datetime | None:
    if last is None or last.submitted_at is None:
        return None
    at = last.submitted_at + timedelta(hours=exam.retry_after_hours)
    return at if at > now else None


def active_attempt(db: Session, user_id: str, exam_id: str, now: datetime) -> ExamAttempt | None:
    """Tentative démarrée, non soumise et encore dans le temps imparti (reprise après fermeture de l'app)."""
    rows = db.scalars(
        select(ExamAttempt)
        .where(ExamAttempt.user_id == user_id, ExamAttempt.exam_id == exam_id, ExamAttempt.submitted_at.is_(None))
        .order_by(ExamAttempt.started_at.desc())
    )
    for row in rows:
        if row.expires_at is not None and row.expires_at > now:
            return row
    return None


def expires_at(exam: ExamSpec, started_at: datetime) -> datetime:
    """Heure limite annoncée au client (sans la grâce, appliquée à la soumission)."""
    return started_at + timedelta(minutes=exam.duration_minutes)


def is_expired(attempt: ExamAttempt, now: datetime) -> bool:
    return attempt.expires_at is not None and now > attempt.expires_at + GRACE


# --- Notation -------------------------------------------------------------------------------


@dataclass(frozen=True)
class GradeResult:
    passed: bool
    global_score: float
    scores: dict[str, float | None]
    gaps: list[dict[str, Any]]
    graded_items: int = 0


def graded_refs(pack: Pack, exam: ExamSpec, index: frozenset[str] | set[str]) -> set[tuple[str, int]]:
    """Items notés compte tenu des médias présents (§1)."""
    concepts = concept_documents(pack)
    return {
        (section["skill"], i)
        for section in exam.sections
        for i, item in enumerate(section["items"])
        if exam_item_graded(item["step"], concepts, index)
    }


def unavailable_reason(pack: Pack, exam: ExamSpec, index: frozenset[str] | set[str]) -> str | None:
    return MEDIA_MISSING if len(graded_refs(pack, exam, index)) < MIN_GRADED_ITEMS else None


def build_items(pack: Pack, exam: ExamSpec, seed: str) -> dict[tuple[str, int], Exercise]:
    """Tous les exercices de la tentative ; ContentError si le contenu est incohérent."""
    content = content_index(pack)
    concepts = lesson_concepts_for_units(pack, exam.requires_units)
    steps = [item["step"] for section in exam.sections for item in section["items"]]
    refs = exam.items()
    return {
        ref: build_exercise(content, exam.id, concepts, steps, flat_index, seed) for flat_index, ref in enumerate(refs)
    }


def item_correct(exercise: Exercise, response: dict[str, Any] | None) -> bool:
    """Réponse manquante, non notée (micro refusé, type inattendu) ou fausse → item faux."""
    if response is None:
        return False
    if exercise.type in SPEECH_TYPES:
        score = response.get("score") if response.get("kind") == "speech" else None
        if not isinstance(score, int | float) or isinstance(score, bool):
            return False
        # Score calculé localement : on lui fait confiance, borné à 0..100.
        return min(100.0, max(0.0, float(score))) >= SPEAK_PASS_SCORE
    result = evaluate(exercise, response)
    return result.graded and result.correct


def grade(
    pack: Pack,
    exam: ExamSpec,
    seed: str,
    answers: list[dict[str, Any]],
    index: frozenset[str] | set[str] | None = None,
) -> GradeResult:
    """Note une tentative ; `index` = médias présents (défaut : ceux du pack publié)."""
    exercises = build_items(pack, exam, seed)
    graded = graded_refs(pack, exam, pack.media_index if index is None else index)
    by_item: dict[tuple[str, int], dict[str, Any]] = {}
    for answer in answers:
        by_item[(answer["section"], answer["index"])] = answer["response"]  # la dernière réponse compte

    totals = dict.fromkeys(SKILLS, 0)
    rights = dict.fromkeys(SKILLS, 0)
    wrong_concepts: dict[str, dict[str, None]] = {}
    for (skill, item_index), exercise in exercises.items():
        if (skill, item_index) not in graded:
            continue  # média manquant : item non noté, hors dénominateur
        totals[skill] += 1
        if item_correct(exercise, by_item.get((skill, item_index))):
            rights[skill] += 1
        else:
            bucket = wrong_concepts.setdefault(skill, {})
            for cid in exercise.concept_ids:
                bucket.setdefault(cid, None)

    scores: dict[str, float | None] = {
        skill: rights[skill] / totals[skill] if totals[skill] else None for skill in SKILLS
    }
    total_items = sum(totals.values())
    global_score = sum(rights.values()) / total_items if total_items else 0.0
    passed = (
        total_items > 0
        and global_score >= exam.pass_threshold - EPSILON
        and all((score or 0.0) >= MIN_SKILL_SCORE - EPSILON for score in scores.values() if score is not None)
    )
    gaps = [{"skill": skill, "conceptIds": list(wrong_concepts[skill])} for skill in SKILLS if skill in wrong_concepts]
    return GradeResult(passed=passed, global_score=global_score, scores=scores, gaps=gaps, graded_items=total_items)


__all__ = [
    "SKILLS",
    "ContentError",
    "ExamSpec",
    "GradeResult",
    "active_attempt",
    "build_items",
    "expires_at",
    "grade",
    "graded_refs",
    "is_expired",
    "is_unlocked",
    "last_submitted",
    "load_exams",
    "next_attempt_at",
    "parse_exam",
    "unavailable_reason",
]

File: packages/core/src/exams.ts
import { checkLessonStep, type ContentIssue } from "./content-checks.ts";
import { buildExercise, evaluate, SPEAK_PASS_SCORE, type Exercise, type ExerciseResponse } from "./engine.ts";
import type { ConceptId, ContentIndex, Curriculum, Lesson, LessonId, LessonStep, Localized, StepType, UnitId } from "./types.ts";

/**
 * Examens de certificat (spec §5.5, docs/contracts/phase2.md §2).
 *
 * Un examen est un fichier de contenu : 4 sections (compétences) d'étapes de leçon.
 * Pour construire les exercices, on fabrique une leçon synthétique dont les étapes
 * sont les items dans l'ordre des sections : l'item n° k (tous items confondus)
 * est l'étape k, et l'ordre des options vient de buildExercise(…, graine) —
 * exactement comme côté serveur, qui rejoue `evaluate` avec la graine `attemptId`.
 */

export const EXAM_SKILLS = ["listening", "reading", "vocabulary", "speaking"] as const;
export type ExamSkill = (typeof EXAM_SKILLS)[number];
export type ExamLevel = "A0" | "A1" | "A2";

export const EXAM_ITEM_COUNT = 25;
export const EXAM_MIN_SECTION_ITEMS = 4;
/** Chaque compétence doit atteindre au moins ce score pour réussir. */
export const EXAM_MIN_SKILL_SCORE = 0.5;
/** Grâce serveur après la durée officielle (contrat §2.3). */
export const EXAM_GRACE_MINUTES = 2;

export type ExamStep = Exclude<LessonStep, { type: "culture_card" | "game" }>;

export interface ExamItem {
  step: ExamStep;
  silent?: true;
}

export interface ExamSection {
  skill: ExamSkill;
  items: ExamItem[];
}

export interface ExamFile {
  id: string;
  level: ExamLevel;
  certificate: Localized;
  requiresUnits: UnitId[];
  durationMinutes: number;
  passThreshold: number;
  retryAfterHours: number;
  sections: ExamSection[];
  reviewed: boolean;
}

/** Référence d'un item : compétence + rang dans la section (forme de l'API). */
export interface ExamItemRef {
  section: ExamSkill;
  index: number;
}

export interface ExamQuestion extends ExamItemRef {
  /** Rang tous items confondus = index d'étape de la leçon synthétique. */
  flatIndex: number;
  stepType: StepType;
  silent: boolean;
  exercise: Exercise;
}

export interface ExamAnswer extends ExamItemRef {
  response: ExerciseResponse;
  responseMs: number;
}

export interface ExamItemResult extends ExamItemRef {
  correct: boolean;
  /** false : item non noté (oral sans micro, examen blanc seulement). */
  graded: boolean;
  conceptIds: ConceptId[];
}

export interface ExamGap {
  skill: ExamSkill;
  conceptIds: ConceptId[];
}

export type ExamScores = Record<ExamSkill, number | null>;

export interface ExamGrade {
  passed: boolean;
  global: number;
  scores: ExamScores;
  gaps: ExamGap[];
  items: ExamItemResult[];
}

const SPEECH_STEPS: ReadonlySet<StepType> = new Set(["speak_repeat", "tone_produce"]);

/** Types d'étape attendus par section (contrat §2.1). Hors liste : avertissement. */
export const EXAM_SECTION_STEPS: Record<ExamSkill, ReadonlySet<StepType>> = {
  listening: new Set(["listen_pick_image", "listen_pick_text", "tone_identify", "tone_minimal_pair"]),
  reading: new Set(["spot_the_south", "build_sentence", "listen_pick_text"]),
  vocabulary: new Set(["listen_pick_image", "build_sentence"]),
  speaking: new Set(["speak_repeat", "tone_produce"]),
};

// ---------------------------------------------------------------------------
// Construction

export function examItemRefs(exam: ExamFile): ExamItemRef[] {
  return exam.sections.flatMap((s) => s.items.map((_, index) => ({ section: s.skill, index })));
}

export function flatIndexOf(exam: ExamFile, ref: ExamItemRef): number {
  let offset = 0;
  for (const s of exam.sections) {
    if (s.skill === ref.section) return ref.index >= 0 && ref.index < s.items.length ? offset + ref.index : -1;
    offset += s.items.length;
  }
  return -1;
}

/** Leçons des unités requises, dans l'ordre du cursus. */
export function examLessons(content: ContentIndex, exam: ExamFile): Lesson[] {
  return exam.requiresUnits.flatMap((unitId) => {
    const unit = content.curriculum.units.find((u) => u.id === unitId);
    return (unit?.lessons ?? []).flatMap((id) => content.lessons.get(id) ?? []);
  });
}

/** Concepts des unités requises (union ordonnée). */
export function examConcepts(content: ContentIndex, exam: ExamFile): ConceptId[] {
  return [...new Set(examLessons(content, exam).flatMap((l) => l.concepts))];
}

/** Leçon synthétique : une étape par item, dans l'ordre des sections. */
export function examLesson(content: ContentIndex, exam: ExamFile): Lesson {
  return {
    id: exam.id,
    unit: exam.requiresUnits[0] ?? "",
    kind: "unit_test",
    title: exam.certificate,
    goal: exam.certificate,
    estimatedMinutes: exam.durationMinutes,
    prerequisites: [],
    concepts: examConcepts(content, exam),
    steps: exam.sections.flatMap((s) => s.items.map((i) => i.step)),
    review: { srsIntroduce: [] },
    reviewed: exam.reviewed,
  };
}

/**
 * Exercices de l'examen pour une graine (attemptId pour l'examen certifiant).
 * `refs` : ordre de passage renvoyé par POST /exams/{id}/start (par défaut, l'ordre du fichier).
 */
export function buildExam(content: ContentIndex, exam: ExamFile, seed: string, refs: readonly ExamItemRef[] = examItemRefs(exam)): ExamQuestion[] {
  const lesson = examLesson(content, exam);
  return refs.flatMap((ref) => {
    const flatIndex = flatIndexOf(exam, ref);
    const item = exam.sections.find((s) => s.skill === ref.section)?.items[ref.index];
    if (flatIndex < 0 || !item) return [];
    return [{ ...ref, flatIndex, stepType: item.step.type, silent: item.silent === true, exercise: buildExercise(content, lesson, flatIndex, seed) }];
  });
}

// ---------------------------------------------------------------------------
// Notation (§2.2) — identique au serveur

export interface GradeOptions {
  /** Examen blanc : un oral sans score (micro refusé) n'est pas noté au lieu d'être faux. */
  allowUngradedSpeech?: boolean;
}

export function gradeItem(question: ExamQuestion, answer: ExamAnswer | undefined, options: GradeOptions = {}): ExamItemResult {
  const base = { section: question.section, index: question.index, conceptIds: question.exercise.conceptIds };
  if (!answer) return { ...base, correct: false, graded: true };
  const { response } = answer;

  if (SPEECH_STEPS.has(question.stepType)) {
    if (response.kind === "speech" && response.score !== null) return { ...base, correct: response.score >= SPEAK_PASS_SCORE, graded: true };
    if (response.kind === "speech" && options.allowUngradedSpeech) return { ...base, correct: false, graded: false };
    return { ...base, correct: false, graded: true };
  }

  const evaluation = evaluate(question.exercise, response);
  return { ...base, correct: evaluation.graded && evaluation.correct, graded: true };
}

const EPSILON = 1e-9;

export function gradeExam(exam: ExamFile, questions: readonly ExamQuestion[], answers: readonly ExamAnswer[], options: GradeOptions = {}): ExamGrade {
  const byRef = new Map(answers.map((a) => [`${a.section}:${a.index}`, a]));
  const items = questions.map((q) => gradeItem(q, byRef.get(`${q.section}:${q.index}`), options));

  const scores = {} as ExamScores;
  let correct = 0;
  let graded = 0;
  const gaps: ExamGap[] = [];
  for (const skill of EXAM_SKILLS) {
    const mine = items.filter((i) => i.section === skill && i.graded);
    const ok = mine.filter((i) => i.correct).length;
    scores[skill] = mine.length === 0 ? null : ok / mine.length;
    correct += ok;
    graded += mine.length;
    const wrong = mine.filter((i) => !i.correct);
    if (wrong.length > 0) gaps.push({ skill, conceptIds: [...new Set(wrong.flatMap((i) => i.conceptIds))] });
  }

  const global = graded === 0 ? 0 : correct / graded;
  const skillsOk = EXAM_SKILLS.every((s) => scores[s] === null ? options.allowUngradedSpeech === true && s === "speaking" : (scores[s] ?? 0) >= EXAM_MIN_SKILL_SCORE - EPSILON);
  return { passed: graded > 0 && global >= exam.passThreshold - EPSILON && skillsOk, global, scores, gaps, items };
}

// ---------------------------------------------------------------------------
// Accès, lacunes

/**
 * Examen débloqué : le test d'unité (`kind: "unit_test"`) de chaque unité requise est fait — terminé,
 * ou sauté grâce au placement si `done` inclut les leçons débloquées. Sans test d'unité (ou sans
 * `lessons`), toutes les leçons de l'unité sont exigées. Même règle que l'API (services/exams.py).
 */
export function isExamUnlocked(
  curriculum: Curriculum,
  exam: ExamFile,
  done: ReadonlySet<LessonId>,
  lessons?: ReadonlyMap<LessonId, Lesson>,
): boolean {
  return exam.requiresUnits.every((unitId) => {
    const unit = curriculum.units.find((u) => u.id === unitId);
    if (!unit || unit.lessons.length === 0) return false;
    const tests = lessons ? unit.lessons.filter((id) => lessons.get(id)?.kind === "unit_test") : [];
    const required = tests.length > 0 ? tests : unit.lessons;
    return required.every((id) => done.has(id));
  });
}

/** Prochaine tentative possible après une tentative soumise (null = tout de suite). */
export function nextExamAttemptAt(exam: ExamFile, lastSubmittedAt: string | null, now: Date = new Date()): Date | null {
  if (!lastSubmittedAt) return null;
  const at = new Date(Date.parse(lastSubmittedAt) + exam.retryAfterHours * 3_600_000);
  return at.getTime() > now.getTime() ? at : null;
}

/**
 * Leçons à revoir pour des concepts manqués : la première leçon (ordre du cursus)
 * qui introduit chaque concept, sinon la première qui l'utilise.
 */
export function lessonsForConcepts(content: ContentIndex, exam: ExamFile, conceptIds: readonly ConceptId[]): LessonId[] {
  const lessons = examLessons(content, exam);
  const found = new Set<LessonId>();
  for (const id of conceptIds) {
    const lesson = lessons.find((l) => l.review.srsIntroduce.includes(id)) ?? lessons.find((l) => l.concepts.includes(id));
    if (lesson) found.add(lesson.id);
  }
  return lessons.filter((l) => found.has(l.id)).map((l) => l.id);
}

/** Leçons à revoir après un examen blanc (items sans concept, ex. spot_the_south, compris). */
export function revisionLessons(content: ContentIndex, exam: ExamFile, questions: readonly ExamQuestion[], grade: ExamGrade): LessonId[] {
  const wrong = grade.items.filter((i) => i.graded && !i.correct);
  const ids = new Set(lessonsForConcepts(content, exam, wrong.flatMap((i) => i.conceptIds)));
  const lessons = examLessons(content, exam);
  for (const item of wrong) {
    const q = questions.find((x) => x.section === item.section && x.index === item.index);
    if (q?.exercise.type === "spot_the_south") {
      const variant = q.exercise.entry.id;
      const lesson = lessons.find((l) => l.steps.some((s) => s.type === "spot_the_south" && s.variant === variant));
      if (lesson) ids.add(lesson.id);
    }
  }
  return lessons.filter((l) => ids.has(l.id)).map((l) => l.id);
}

// ---------------------------------------------------------------------------
// Validation sémantique (scripts/validate-content.ts)

function stepConcepts(step: ExamStep): ConceptId[] {
  switch (step.type) {
    case "listen_pick_image":
      return [step.concept, ...step.distractors];
    case "listen_pick_text":
    case "listen_transcribe":
    case "tone_identify":
    case "tone_produce":
    case "speak_repeat":
      return [step.concept];
    case "tone_minimal_pair":
      return step.audioConcepts ?? [];
    case "build_sentence":
      return step.audioConcept ? [step.audioConcept] : [];
    case "match_pairs":
      return step.concepts;
    default:
      return [];
  }
}

export function checkExam(content: ContentIndex, exam: ExamFile, where = exam.id): ContentIssue[] {
  const issues: ContentIssue[] = [];
  const err = (message: string, at = where) => issues.push({ level: "error", where: at, message });
  const warn = (message: string, at = where) => issues.push({ level: "warning", where: at, message });

  for (const unitId of exam.requiresUnits) {
    const unit = content.curriculum.units.find((u) => u.id === unitId);
    if (!unit) err(`Unité requise inconnue : ${unitId}`);
    else if (unit.lessons.length === 0) err(`Unité requise sans leçon : ${unitId}`);
  }

  const total = exam.sections.reduce((n, s) => n + s.items.length, 0);
  if (total !== EXAM_ITEM_COUNT) err(`${total} items (attendu : ${EXAM_ITEM_COUNT})`);
  for (const skill of EXAM_SKILLS) {
    const count = exam.sections.filter((s) => s.skill === skill).length;
    if (count !== 1) err(`Section « ${skill} » présente ${count} fois (attendu : 1)`);
  }

  const lesson = examLesson(content, exam);
  const allowed = new Set(lesson.concepts);
  for (const section of exam.sections) {
    if (section.items.length < EXAM_MIN_SECTION_ITEMS) err(`Section ${section.skill} : ${section.items.length} items (minimum ${EXAM_MIN_SECTION_ITEMS})`);
    section.items.forEach((item, i) => {
      const at = `${where} ${section.skill}[${i}] (${item.step.type})`;
      if (!EXAM_SECTION_STEPS[section.skill].has(item.step.type)) warn(`Type d'étape inattendu dans la section ${section.skill}`, at);
      if (item.silent && (section.skill === "listening" || section.skill === "speaking")) err(`silent interdit en ${section.skill}`, at);
      for (const id of stepConcepts(item.step)) {
        if (!allowed.has(id)) err(`Concept ${id} hors des unités requises`, at);
      }
      for (const issue of checkLessonStep(content, lesson, item.step, at)) {
        // Les concepts sont contrôlés ci-dessus contre les unités requises.
        if (!issue.message.includes("absent de lesson.concepts")) issues.push(issue);
      }
    });
  }

  if (!issues.some((i) => i.level === "error")) {
    try {
      buildExam(content, exam, "validation");
    } catch (error) {
      err(`Construction impossible : ${error instanceof Error ? error.message : String(error)}`);
    }
  }
  if (!exam.reviewed) warn(`Examen non relu par un locuteur natif`);
  return issues;
}
